### PR TITLE
Fix and test permonst deallocation issue

### PR DESCRIPTION
**Pull request by heiner**

### CODE REVIEW

File: nle/tests/test_nethack.py [hunk 1]
```diff
@@ -89,6 +89,21 @@ def test_permonst(self):
...
+            self.assertEqual(mon.mname, "giant ant")
+
+    def test_illegal_numbers(self):
+        with self.assertRaisesRegex(IndexError, "NUMMONS"):
+            nethack.permonst(10000)
+
+        with self.assertRaisesRegex(IndexError, "NUMMONS"):
+            nethack.permonst(-1)
+
+        with self.assertRaisesRegex(IndexError, "MAXMCLASSES"):
```
Comment (edran) [inline nle/tests/test_nethack.py:104]: Just to be consistent, I would probably add an exception check for negative values for `from_mlet`.

File: nle/tests/test_nethack.py [hunk 2]
```diff
@@ -89,6 +89,21 @@ def test_permonst(self):
...
         self.assertEqual(mon.mname, "killer bee")
 
+        # Test https://github.com/pybind/pybind11/issues/2394
+        for _ in range(2):
+            mon = nethack.permonst(0)
+            self.assertEqual(mon.mname, "giant ant")
+
+    def test_illegal_numbers(self):
+        with self.assertRaisesRegex(IndexError, "NUMMONS"):
+            nethack.permonst(10000)
```
Comment (edran) [inline nle/tests/test_nethack.py:99]: Do (or should?) we have access to `NUMMONS` at runtime here? That way we could check for `upper_bound + 1`.

EDIT: better question

**Review by edran — APPROVED** [review]
LGTM! Thanks, well spotted!

File: win/rl/helper.cc [hunk 3]
```diff
@@ -124,9 +124,18 @@ PYBIND11_MODULE(helper, m)
...
-    py::class_<permonst, std::unique_ptr<permonst, py::nodelete> >(m,
-                                                                   "permonst")
-        .def(py::init([](int index) -> permonst * { return &mons[index]; }))
+    py::class_<permonst>(m, "permonst")
+        .def(
+            "__init__",
+            [](py::detail::value_and_holder &v_h, int index) {
+                if (index < 0 || index >= NUMMONS)
+                    throw std::out_of_range(
+                        "Index should be between 0 and NUMMONS");
```
Comment (alexholdenmiller) [inline win/rl/helper.cc:133]: should the error include what NUMMONS is, e.g. "Index should be between 0 and NUMMONS (int)"

File: win/rl/helper.cc [hunk 4]
```diff
@@ -159,7 +168,12 @@ PYBIND11_MODULE(helper, m)
     py::class_<class_sym>(m, "class_sym")
         .def_static(
             "from_mlet",
-            [](char let) -> const class_sym * { return &def_monsyms[let]; },
+            [](char let) -> const class_sym * {
+                if (let < 0 || let >= MAXMCLASSES)
+                    throw std::out_of_range(
+                        "Argument should be between 0 and MAXMCLASSES");
```
Comment (alexholdenmiller) [inline win/rl/helper.cc:174]: same as above

File: nle/tests/test_nethack.py (hunk 1 above)
Comment (heiner) [reply nle/tests/test_nethack.py:104]: I would if I could -- but mlet is a char which in Python gets mapped to a string of length 1. I cannot produce negative values there :D

File: nle/tests/test_nethack.py (hunk 2 above)
Comment (heiner) [reply nle/tests/test_nethack.py]: Done.

File: win/rl/helper.cc (hunk 3 above)
Comment (heiner) [reply win/rl/helper.cc]: Kinda weird but okay.

Comment (heiner) [reply win/rl/helper.cc]: Ah I begin to understand what you meant. Fixed again.

**Pull request merged**